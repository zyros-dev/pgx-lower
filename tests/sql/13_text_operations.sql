-- Test text operations: PgTextAddOp, PgConcatOp, PgLikeOp, PgSubstringOp
LOAD 'pgx_lower';

DROP TABLE IF EXISTS test_text;

CREATE TABLE test_text (
    id SERIAL PRIMARY KEY,
    name TEXT,
    description TEXT
);

INSERT INTO test_text(name, description) VALUES 
    ('Alice', 'Software Engineer'),
    ('Bob', 'Data Scientist'),
    ('Charlie', 'Product Manager'),
    ('Diana', 'UX Designer');

-- Test text operations on table columns
-- These should trigger MLIR compilation with text operators
SELECT id, name, description FROM test_text;

DROP TABLE test_text;