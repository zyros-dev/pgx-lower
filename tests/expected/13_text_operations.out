-- Test text operations: PgTextAddOp, PgConcatOp, PgLikeOp, PgSubstringOp
LOAD 'pgx_lower';
NOTICE:  Installing custom executor hook...
NOTICE:  Registering custom sigsegv handler!
DROP TABLE IF EXISTS test_text;
NOTICE:  table "test_text" does not exist, skipping
CREATE TABLE test_text (
    id SERIAL PRIMARY KEY,
    name TEXT,
    description TEXT
);
INSERT INTO test_text(name, description) VALUES 
    ('Alice', 'Software Engineer'),
    ('Bob', 'Data Scientist'),
    ('Charlie', 'Product Manager'),
    ('Diana', 'UX Designer');
NOTICE:  Custom executor is being executed in C!
NOTICE:  Calling C++ executor from C...
NOTICE:  Successfully created MLIR context!
NOTICE:  LLVM version: 20.1.2
NOTICE:  Inside C++ executor! Plan type: 3
NOTICE:  Query text: INSERT INTO test_text(name, description) VALUES 
    ('Alice', 'Software Engineer'),
    ('Bob', 'Data Scientist'),
    ('Charlie', 'Product Manager'),
    ('Diana', 'UX Designer');
NOTICE:  Not a SELECT statement, skipping
NOTICE:  MLIR couldn't handle query, falling back to standard executor
-- Test text operations on table columns
-- These should trigger MLIR compilation with text operators
SELECT id, name, description FROM test_text;
NOTICE:  Custom executor is being executed in C!
NOTICE:  Calling C++ executor from C...
NOTICE:  Successfully created MLIR context!
NOTICE:  LLVM version: 20.1.2
NOTICE:  Inside C++ executor! Plan type: 1
NOTICE:  Query text: SELECT id, name, description FROM test_text;
NOTICE:  Query analysis: Simple sequential scan - MLIR compatible
NOTICE:  Table has 2 unsupported column types out of 3 total - MLIR will attempt to handle with fallback values
NOTICE:  Generated MLIR with PostgreSQL typed field access:
NOTICE:  MLIR with field access: module {
  func.func private @open_postgres_table(i64) -> i64
  func.func private @read_next_tuple_from_table(i64) -> i64
  func.func private @close_postgres_table(i64)
  func.func private @get_int_field(i64, i32, !llvm.ptr) -> i32
  func.func private @get_text_field(i64, i32, !llvm.ptr) -> i64
  func.func private @add_tuple_to_result(i64) -> i1
  func.func @main() -> i64 {
    %0 = pg.scan_table "current_table" : !pg.table_handle
    %c0_i64 = arith.constant 0 : i64
    %c-2_i64 = arith.constant -2 : i64
    %true = arith.constant true
    %1:2 = scf.while (%arg0 = %true, %arg1 = %c0_i64) : (i1, i64) -> (i1, i64) {
      scf.condition(%arg0) %arg0, %arg1 : i1, i64
    } do {
    ^bb0(%arg0: i1, %arg1: i64):
      %2 = pg.read_tuple %0 : !pg.table_handle -> !pg.tuple_handle
      %3 = builtin.unrealized_conversion_cast %2 : !pg.tuple_handle to i64
      %4 = arith.cmpi eq, %3, %c-2_i64 : i64
      %5:2 = scf.if %4 -> (i1, i64) {
        %false = arith.constant false
        scf.yield %false, %arg1 : i1, i64
      } else {
        %true_1 = arith.constant true
        %6 = func.call @add_tuple_to_result(%3) : (i64) -> i1
        %c1_i64 = arith.constant 1 : i64
        %7 = arith.addi %arg1, %c1_i64 : i64
        scf.yield %true_1, %7 : i1, i64
      }
      scf.yield %5#0, %5#1 : i1, i64
    }
    %c0_i64_0 = arith.constant 0 : i64
    call @close_postgres_table(%c0_i64_0) : (i64) -> ()
    return %1#1 : i64
  }
}

NOTICE:  Applied pg-to-scf lowering pass!
NOTICE:  Lowered MLIR: module {
  func.func private @open_postgres_table(i64) -> i64
  func.func private @read_next_tuple_from_table(i64) -> i64
  func.func private @close_postgres_table(i64)
  func.func private @get_int_field(i64, i32, !llvm.ptr) -> i32
  func.func private @get_text_field(i64, i32, !llvm.ptr) -> i64
  func.func private @add_tuple_to_result(i64) -> i1
  func.func @main() -> i64 {
    %c1_i64 = arith.constant 1 : i64
    %false = arith.constant false
    %true = arith.constant true
    %c-2_i64 = arith.constant -2 : i64
    %c0_i64 = arith.constant 0 : i64
    %c-5956942909115520617_i64 = arith.constant -5956942909115520617 : i64
    %0 = call @open_postgres_table(%c-5956942909115520617_i64) : (i64) -> i64
    %1:2 = scf.while (%arg0 = %true, %arg1 = %c0_i64) : (i1, i64) -> (i1, i64) {
      scf.condition(%arg0) %arg0, %arg1 : i1, i64
    } do {
    ^bb0(%arg0: i1, %arg1: i64):
      %2 = func.call @read_next_tuple_from_table(%0) : (i64) -> i64
      %3 = arith.cmpi eq, %2, %c-2_i64 : i64
      %4:2 = scf.if %3 -> (i1, i64) {
        scf.yield %false, %arg1 : i1, i64
      } else {
        %5 = func.call @add_tuple_to_result(%2) : (i64) -> i1
        %6 = arith.addi %arg1, %c1_i64 : i64
        scf.yield %true, %6 : i1, i64
      }
      scf.yield %4#0, %4#1 : i1, i64
    }
    call @close_postgres_table(%c0_i64) : (i64) -> ()
    return %1#1 : i64
  }
}

NOTICE:  Lowered PostgreSQL typed field access MLIR to LLVM dialect!
NOTICE:  Created MLIR ExecutionEngine for PostgreSQL typed field access!
NOTICE:  Invoked MLIR JIT PostgreSQL typed field access!
NOTICE:  PostgreSQL typed field access completed with result: 4
NOTICE:  MLIR successfully handled the query
NOTICE:  MLIR successfully handled the query
 id |  name   |    description    
----+---------+-------------------
  1 | Alice   | Software Engineer
  2 | Bob     | Data Scientist
  3 | Charlie | Product Manager
  4 | Diana   | UX Designer
(4 rows)

DROP TABLE test_text;
